spec/validations/requirements_spec.rb

```diff
@@ -194,48 +223,54 @@
     end
   end
 
-  context 'there is a valid custom resources schema defined' do
-    let(:requirements_string) do
-      JSON.generate(
-        'custom_resources_schema' => { 'resource_types' => [], 'relationship_types' => [] }
-      )
-    end
+  context 'custom object requirements validations' do
+    context 'there is a valid custom objects schema defined' do
+      let(:requirements_string) do
+        JSON.generate(
+          'custom_objects' => {
+            'custom_object_types' => [
+              'key' => 'foo',
+              'schema' => {}
+            ],
+            'custom_object_relationship_types' => []
+          }
+        )
+      end
 
-    it 'does not return an error' do
-      expect(errors).to be_empty
+      it 'does not return an error' do
+        expect(errors).to be_empty
+      end
     end
-  end
 
-  context 'a custom resources schema contains invalid keys' do
-    let(:requirements_string) do
-      JSON.generate(
-        'custom_resources_schema' => {
-          'resource_types' => [], 'relationship_types' => [], 'resources' => [], 'relationships' => []
-        }
-      )
-    end
+    context 'a custom objects schema contains invalid type-level keys' do
+      let(:requirements_string) do
+        JSON.generate(
+          'custom_objects' => { 'custom_object_types' => [{}], 'custom_object_relationship_types' => [] }
+        )
+      end
 
-    it 'creates an error' do
-      expect(errors.first.key).to eq(:invalid_cr_schema_keys)
-      expect(errors.first.data).to eq(invalid_keys: 'resources, relationships', count: 2)
+      it 'creates an error' do
+        expect(errors.first.key).to eq(:missing_required_fields)
+        expect(errors.first.data).to eq(field: 'key', identifier: 'custom_object_types')
```

Comment:
Instead of `errors.first.data` or `errors.first.key`. Would it be better to test that
ValidationError.new to be called with`:missing_required_fields , field: ..., identifier: ...`

```
it 'creates an error' do
  expect(ValidationError).to receive(:new).with(:missing_required_fields , field: 'key', identifier: 'custom_object_types')
end
```

Reply:
I would say no for two reasons

The first is that a `ValidationError` being initialised doesn't necessarily mean it was added to the errors. Imagine a case where there has been a refactor to make this more functional and have the `ValidationError` object be returned from a method, but the definition of the parent method calling that forgets to push it onto the errors object. The test would pass, even though the validation was not performed correctly.

The second is that this would couple the test more closely with the implementation. Currently, the test has no knowledge of the `ValidationError` class, it only focuses that the output has the correct data in the correct structure. Should this class be renamed, moved, replaced or deleted, this test wouldn't care. As long as the output is what was expected. ie an array with the first object responding to `key` with the expected key and `data` with the expected hash

Reply:
I feel you. 👐 

>Should this class be renamed, moved, replaced or deleted, this test wouldn't care. As long as the output is what was expected. ie an array with the first object responding to key with the expected key and data with the expected hash

This is like duck typing. as long as it quacks it is a duck. As long as it responds to `.first`, `.first.key` and `first.data`, it is an array containing an object with those methods.

I also understand your view on not coupling the test closely. When it is coupled and the class knows about ValidationError, it makes future change and improvements challenging. With this being a unit test, I always forget to test in isolation. I will be more deliberate moving forward. 

Thank you for the insight 👍